File: src/quest_editor/actions/EditEntityPropAction.ts
import { Action } from "../../core/undo/Action";
import { QuestEntityPropModel } from "../model/QuestEntityPropModel";
import { QuestEditorStore } from "../stores/QuestEditorStore";
import { QuestEntityModel } from "../model/QuestEntityModel";

export class EditEntityPropAction implements Action {
    readonly description: string;

    constructor(
        private readonly quest_editor_store: QuestEditorStore,
        private readonly entity: QuestEntityModel,
        private readonly prop: QuestEntityPropModel,
        private readonly old_value: number,
        private readonly new_value: number,
    ) {
        this.description = `Edit ${prop.name}`;
    }

    redo(): void {
        this.prop.set_value(this.new_value);
        this.quest_editor_store.set_selected_entity(this.entity);
    }

    undo(): void {
        this.prop.set_value(this.old_value);
        this.quest_editor_store.set_selected_entity(this.entity);
    }
}

File: web/src/main/kotlin/world/phantasmal/web/questEditor/actions/EditEntityPropAction.kt
package world.phantasmal.web.questEditor.actions

import world.phantasmal.observable.change
import world.phantasmal.web.core.actions.Action
import world.phantasmal.web.questEditor.models.QuestEntityModel
import world.phantasmal.web.questEditor.models.QuestEntityPropModel

class EditEntityPropAction(
    private val setSelectedEntity: (QuestEntityModel<*, *>) -> Unit,
    private val entity: QuestEntityModel<*, *>,
    private val prop: QuestEntityPropModel,
    private val newValue: Any,
    private val oldValue: Any,
) : Action {
    override val description: String = "Edit ${entity.type.simpleName} ${prop.name}"

    override fun execute() {
        change {
            setSelectedEntity(entity)
            prop.setValue(newValue)
        }
    }

    override fun undo() {
        change {
            setSelectedEntity(entity)
            prop.setValue(oldValue)
        }
    }
}
